Python/bytecodes.c

```diff
@@ -409,104 +409,134 @@ dummy_func(
-            _GUARD_BOTH_INT + unused/1 + _BINARY_OP_MULTIPLY_INT;
-        macro(BINARY_OP_ADD_INT) =
-            _GUARD_BOTH_INT + unused/1 + _BINARY_OP_ADD_INT;
-        macro(BINARY_OP_SUBTRACT_INT) =
-            _GUARD_BOTH_INT + unused/1 + _BINARY_OP_SUBTRACT_INT;
-
-        op(_GUARD_BOTH_FLOAT, (left, right -- left, right)) {
-            EXIT_IF(!PyFloat_CheckExact(left));
-            EXIT_IF(!PyFloat_CheckExact(right));
+        op(_BINARY_OP_TABLE_ND, (type_version/1, left, right -- res)) {
+            PyTypeObject *lt = Py_TYPE(left);
+            PyTypeObject *rt = Py_TYPE(right);
+            EXIT_IF(lt->tp_version_tag != ((type_version & 0xf0) >> 4));
+            EXIT_IF(rt->tp_version_tag != (type_version & 0xf));
+            STAT_INC(BINARY_OP, hit);
+            res = _Py_BinaryFunctionTable[type_version >> 8](left, right);
+            Py_DECREF(right);
+            ERROR_IF(res == NULL, error);
         }
 
-        pure op(_BINARY_OP_MULTIPLY_FLOAT, (left, right -- res)) {
+        op(_BINARY_OP_TABLE_DN, (type_version/1, left, right -- res)) {
+            PyTypeObject *lt = Py_TYPE(left);
+            PyTypeObject *rt = Py_TYPE(right);
+            EXIT_IF(lt->tp_version_tag != ((type_version & 0xf0) >> 4));
+            EXIT_IF(rt->tp_version_tag != (type_version & 0xf));
             STAT_INC(BINARY_OP, hit);
-            double dres =
-                ((PyFloatObject *)left)->ob_fval *
-                ((PyFloatObject *)right)->ob_fval;
-            DECREF_INPUTS_AND_REUSE_FLOAT(left, right, dres, res);
+            res = _Py_BinaryFunctionTable[type_version >> 8](left, right);
+            Py_DECREF(left);
+            ERROR_IF(res == NULL, error);
         }
 
-        pure op(_BINARY_OP_ADD_FLOAT, (left, right -- res)) {
+        op(_BINARY_OP_TABLE_DD, (type_version/1, left, right -- res)) {
+            PyTypeObject *lt = Py_TYPE(left);
+            PyTypeObject *rt = Py_TYPE(right);
+            EXIT_IF(lt->tp_version_tag != ((type_version & 0xf0) >> 4));
+            EXIT_IF(rt->tp_version_tag != (type_version & 0xf));
             STAT_INC(BINARY_OP, hit);
-            double dres =
-                ((PyFloatObject *)left)->ob_fval +
-                ((PyFloatObject *)right)->ob_fval;
-            DECREF_INPUTS_AND_REUSE_FLOAT(left, right, dres, res);
+            res = _Py_BinaryFunctionTable[type_version >> 8](left, right);
+            Py_DECREF(left);
+            Py_DECREF(right);
+            ERROR_IF(res == NULL, error);
         }
 
-        pure op(_BINARY_OP_SUBTRACT_FLOAT, (left, right -- res)) {
-            STAT_INC(BINARY_OP, hit);
-            double dres =
-                ((PyFloatObject *)left)->ob_fval -
-                ((PyFloatObject *)right)->ob_fval;
-            DECREF_INPUTS_AND_REUSE_FLOAT(left, right, dres, res);
+        macro(BINARY_OP_1I) =
+            unused/1 +
+            _GUARD_NOS_REFCNT1 +
+            _GUARD_TOS_IMMORTAL +
+            _BINARY_OP_TABLE_NN;
```

Comment:
Can you please document in the `bytecodes.c` file what these suffixes mean? Some of them are easy to see, but I don't know what's `NN` vs `ND` immediately.

Reply:
Done.